tooling/greybox_fuzzer/src/coverage.rs

```diff
@@ -0,0 +1,751 @@
+                        != cmp_approach.encountered_loop_log2s
+                    {
+                        return true;
+                    }
+                    if last_value > cmp_approach.encountered_loop_maximum {
+                        return true;
+                    }
+                }
+                std::cmp::Ordering::Greater => {}
+            }
+        }
+        // Check if a boolean state has been observed before
+        for acir_bool_state in new_coverage.acir_bool_coverage.iter() {
+            if !self.acir_bool_coverage.contains(acir_bool_state) {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+/// Analyze the brillig program to detect:
+/// 1. How many branches and conditional moves there are (needed fro branch coverage)
+/// 2. How many comparisons there are (needed for comparison coverage)
+///
+/// Provide a feature to raw index map for brillig
+pub fn analyze_brillig_program_before_fuzzing(
+    program: &ProgramArtifact,
+) -> (FeatureToIndexMap, BrilligCoverageRanges) {
+    // Brillig program is an ACIR program where the first opcode is brillig call
+    let program_bytecode = &program.bytecode;
+    let main_function = &program_bytecode.functions[0];
+    let starting_opcode = &main_function.opcodes[0];
+
+    let fuzzed_brillig_function_id = match starting_opcode {
+        Opcode::BrilligCall { id, .. } => id,
+        _ => panic!(
+            "If a method is compiled to brillig, the first opcode in ACIR has to be brillig call"
+        ),
+    };
+    // Get the brillig code
+    let fuzzed_brillig_function =
+        &program_bytecode.unconstrained_functions[fuzzed_brillig_function_id.as_usize()];
+
+    let mut feature_to_index_map = HashMap::new();
+    let mut total_features = 0usize;
+    let mut coverage_items = BrilligCoverageRanges::new();
+    // Go through each opcode, detect branching and comparison opcodes and then store information about them
+    for (opcode_index, opcode) in fuzzed_brillig_function.bytecode.iter().enumerate() {
+        match opcode {
+            // Conditional branching
+            &BrilligOpcode::JumpIf { location, .. }
+            | &BrilligOpcode::JumpIfNot { location, .. } => {
+                feature_to_index_map.insert((opcode_index, location), total_features);
+                feature_to_index_map.insert((opcode_index, opcode_index + 1), total_features + 1);
+                coverage_items.push(BrilligCoverageItemRange::Branch(BranchCoverageRange {
+                    index: total_features,
+                }));
+                total_features += 2;
+            }
+            // Conditional mov
+            &BrilligOpcode::ConditionalMov { .. } => {
+                feature_to_index_map.insert((opcode_index, usize::MAX - 1), total_features);
+                feature_to_index_map.insert((opcode_index, usize::MAX), total_features + 1);
```

Comment:
I looked at the tracing code first and couldn't understand why `usize::MAX` and `usize::MAX - 1` were used there, but now seeing this code I understand. I wonder if these could be extracted into documented constants so that it's easier to see where they come from and what they mean.

Reply:
I changed it